File: src/shell/moduleConfig.ts
import type { LucideIcon } from 'lucide-react'
import {
  Sprout,
  Landmark,
  HandCoins,
  ShoppingBag,
  Warehouse,
  LayoutDashboard,
  Activity,
  Wallet,
  MoreHorizontal,
  Home,
  Receipt,
  CreditCard,
  FileText,
  Calculator,
  Boxes,
  ArrowLeftRight,
  ListOrdered,
  LayoutGrid,
  Menu,
  Beef,
  Package,
  Users,
  Search,
  RefreshCw,
  Zap,
  History,
  SlidersHorizontal,
  HelpCircle,
  ClipboardList,
  FileSignature,
  Heart,
  BarChart3,
} from 'lucide-react'

/**
 * Registro central de módulos do superapp (spec §3.4/§7.3).
 * O Shell itera este registro para montar a Barra de Módulos e injeta os `bottomTabs`
 * do módulo ativo no BottomTabBar genérico. Nenhuma navegação de módulo é hardcodada fora daqui.
 */

export interface BottomTab {
  id: string
  label: string
  icon: LucideIcon
  /** rota relativa dentro do módulo (ex.: '' = home, 'atividades'). */
  path: string
  /** botão central elevado (ex.: "Registrar" em campo) */
  elevated?: boolean
  /** ação especial em vez de navegação (ex.: 'menu' abre o RevealMenu global) */
  action?: 'menu'
}

export interface ModuleMenuItem {
  id: string
  label: string
  icon: LucideIcon
  /** rota absoluta do destino */
  route: string
}

export interface ModuleMenuSection {
  title: string
  items: ModuleMenuItem[]
}

export interface ModuleDef {
  id: string
  label: string
  icon: LucideIcon
  /** rota inicial absoluta do módulo */
  homeRoute: string
  bottomTabs: BottomTab[]
  /**
   * funcionalidades exibidas no RevealMenu (aba Mais) — contextuais ao módulo.
   * Quando ausente, o menu deriva uma seção única das bottomTabs navegáveis.
   */
  menuSections?: ModuleMenuSection[]
}

/** Fallback do RevealMenu: seção única derivada das abas navegáveis do módulo. */
export function getMenuSections(module: ModuleDef): ModuleMenuSection[] {
  if (module.menuSections) return module.menuSections
  return [
    {
      title: 'Funcionalidades',
      items: module.bottomTabs
        .filter((tab) => tab.path !== '' && !tab.action)
        .map((tab) => ({ id: tab.id, label: tab.label, icon: tab.icon, route: `/${module.id}/${tab.path}` })),
    },
  ]
}

export const MODULES: ModuleDef[] = [
  {
    // New-UI — hub agregador: porta de entrada do superapp, Banking central
    id: 'inicio',
    label: 'Início',
    icon: Home,
    homeRoute: '/inicio',
    bottomTabs: [
      { id: 'home', label: 'Início', icon: Home, path: '' },
      { id: 'apps', label: 'Apps', icon: LayoutGrid, path: 'apps' },
      { id: 'carteira', label: 'Carteira', icon: Wallet, path: 'carteira' },
      { id: 'menu', label: 'Menu', icon: Menu, path: 'menu', action: 'menu' },
    ],
  },
  {
    id: 'fazendas',
    label: 'Fazendas',
    icon: Sprout,
    homeRoute: '/fazendas',
    bottomTabs: [
      { id: 'dashboard', label: 'Dashboard', icon: LayoutDashboard, path: '' },
      { id: 'fazendas', label: 'Fazendas', icon: Sprout, path: 'fazendas' },
      { id: 'atividades', label: 'Atividades', icon: Activity, path: 'atividades' },
      { id: 'financeiro', label: 'Financeiro', icon: Wallet, path: 'financeiro' },
      { id: 'mais', label: 'Mais', icon: MoreHorizontal, path: 'mais', action: 'menu' },
    ],
    menuSections: [
      {
        title: 'Dashboards gerenciais',
        items: [
          { id: 'financeiro', label: 'Financeiro', icon: Wallet, route: '/fazendas/dashboards/financeiro' },
          { id: 'pecuaria', label: 'Pecuária de Corte', icon: Beef, route: '/fazendas/dashboards/pecuaria' },
          { id: 'confinamento', label: 'Lotação de Currais', icon: Warehouse, route: '/fazendas/dashboards/confinamento' },
          { id: 'ativos', label: 'Ativos / Depreciação', icon: Package, route: '/fazendas/dashboards/ativos' },
          { id: 'suprimentos', label: 'Suprimentos', icon: Boxes, route: '/fazendas/dashboards/suprimentos' },
          { id: 'uso', label: 'Análise de Uso', icon: Users, route: '/fazendas/dashboards/uso' },
          { id: 'consultas', label: 'Consultas Gerenciais', icon: Search, route: '/fazendas/dashboards/consultas' },
        ],
      },
      {
        title: 'Operacional',
        items: [
          { id: 'sync', label: 'Fila de sincronização', icon: RefreshCw, route: '/fazendas/mais/sync' },
          { id: 'atividades', label: 'Todas as atividades', icon: Activity, route: '/fazendas/atividades' },
        ],
      },
    ],
  },
  {
    id: 'bank',
    label: 'Bank',
    icon: Landmark,
    homeRoute: '/bank',
    bottomTabs: [
      { id: 'inicio', label: 'Início', icon: Home, path: '' },
      { id: 'extrato', label: 'Extrato', icon: Receipt, path: 'extrato' },
      { id: 'pagamentos', label: 'Pagamentos', icon: ArrowLeftRight, path: 'pagamentos' },
      { id: 'cartoes', label: 'Cartões', icon: CreditCard, path: 'cartoes' },
      { id: 'mais', label: 'Mais', icon: MoreHorizontal, path: 'mais', action: 'menu' },
    ],
    menuSections: [
      {
        title: 'Pagamentos e transferências',
        items: [
          { id: 'pix', label: 'Pix', icon: Zap, route: '/bank/pix' },
          { id: 'pagamentos', label: 'Pagamentos', icon: Receipt, route: '/bank/pagamentos' },
          { id: 'extrato', label: 'Extrato', icon: History, route: '/bank/extrato' },
        ],
      },
      {
        title: 'Cartão',
        items: [
          { id: 'cartoes', label: 'Cartões', icon: CreditCard, route: '/bank/cartoes' },
          { id: 'limites', label: 'Limites', icon: SlidersHorizontal, route: '/bank/limites' },
        ],
      },
      {
        title: 'Suporte',
        items: [{ id: 'ajuda', label: 'Ajuda', icon: HelpCircle, route: '/bank/ajuda' }],
      },
    ],
  },
  {
    id: 'credito',
    label: 'Crédito',
    icon: HandCoins,
    homeRoute: '/credito',
    bottomTabs: [
      { id: 'inicio', label: 'Início', icon: Home, path: '' },
      { id: 'propostas', label: 'Minhas Propostas', icon: FileText, path: 'propostas' },
      { id: 'simular', label: 'Simular', icon: Calculator, path: 'simular' },
      { id: 'mais', label: 'Mais', icon: MoreHorizontal, path: 'mais', action: 'menu' },
    ],
    menuSections: [
      {
        title: 'Crédito',
        items: [
          { id: 'simular', label: 'Simular', icon: Calculator, route: '/credito/simular' },
          { id: 'propostas', label: 'Propostas', icon: ClipboardList, route: '/credito/propostas' },
          { id: 'contratos', label: 'Contratos', icon: FileSignature, route: '/credito/contratos' },
        ],
      },
      {
        title: 'Suporte',
        items: [{ id: 'ajuda', label: 'Ajuda', icon: HelpCircle, route: '/credito/ajuda' }],
      },
    ],
  },
  {
    id: 'marketplace',
    label: 'Marketplace',
    icon: ShoppingBag,
    homeRoute: '/marketplace',
    bottomTabs: [
      { id: 'inicio', label: 'Início', icon: Home, path: '' },
      { id: 'categorias', label: 'Categorias', icon: ListOrdered, path: 'categorias' },
      { id: 'pedidos', label: 'Pedidos', icon: Receipt, path: 'pedidos' },
      { id: 'mais', label: 'Mais', icon: MoreHorizontal, path: 'mais', action: 'menu' },
    ],
    menuSections: [
      {
        title: 'Compras',
        items: [
          { id: 'categorias', label: 'Categorias', icon: LayoutGrid, route: '/marketplace/categorias' },
          { id: 'pedidos', label: 'Pedidos', icon: Package, route: '/marketplace/pedidos' },
          { id: 'favoritos', label: 'Favoritos', icon: Heart, route: '/marketplace/favoritos' },
        ],
      },
      {
        title: 'Suporte',
        items: [{ id: 'ajuda', label: 'Ajuda', icon: HelpCircle, route: '/marketplace/ajuda' }],
      },
    ],
  },
  {
    id: 'armazem',
    label: 'Armazém',
    icon: Warehouse,
    homeRoute: '/armazem',
    bottomTabs: [
      { id: 'inicio', label: 'Início', icon: Home, path: '' },
      { id: 'estoque', label: 'Estoque', icon: Boxes, path: 'estoque' },
      { id: 'movimentacoes', label: 'Movimentações', icon: ArrowLeftRight, path: 'movimentacoes' },
      { id: 'mais', label: 'Mais', icon: MoreHorizontal, path: 'mais', action: 'menu' },
    ],
    menuSections: [
      {
        title: 'Operação',
        items: [
          { id: 'estoque', label: 'Estoque', icon: Boxes, route: '/armazem/estoque' },
          { id: 'movimentacoes', label: 'Movimentações', icon: ArrowLeftRight, route: '/armazem/movimentacoes' },
          { id: 'unidades', label: 'Unidades', icon: Warehouse, route: '/armazem/unidades' },
        ],
      },
      {
        title: 'Gestão',
        items: [{ id: 'relatorios', label: 'Relatórios', icon: BarChart3, route: '/armazem/relatorios' }],
      },
    ],
  },
]

export const MODULE_MAP: Record<string, ModuleDef> = Object.fromEntries(MODULES.map((m) => [m.id, m]))

export function getModule(id: string | undefined): ModuleDef | undefined {
  return id ? MODULE_MAP[id] : undefined
}

File: apps/mobile/lib/shell/module_config.dart
import '../modules/fazendas/functional_catalog.dart';
import '../modules/fazendas/group_icons.dart';
import '../modules/fazendas/operational_groups.dart';
import '../ui/ui.dart';
import 'state/prototype_session_store.dart';

/// Registro central de navegação — espelha `moduleConfig.ts` (spec §3.4/§7.3).
/// O app tem um módulo (Fazendas). A navbar vem de [operationalBottomTabs] e o
/// menu lateral de [operationalMenuSections]; nenhuma navegação é hardcodada
/// fora daqui.

class BottomTab {
  const BottomTab({
    required this.id,
    required this.label,
    required this.icon,
    required this.path,
    this.elevated = false,
    this.action,
    this.profiles = const {},
  });

  final String id;
  final String label;
  final AppIconData icon;

  /// Rota relativa dentro do módulo (ex.: '' = home, 'atividades').
  final String path;

  /// Botão central elevado (ex.: "Registrar" em campo).
  final bool elevated;

  /// Ação especial em vez de navegação (ex.: 'menu' abre o RevealMenu global).
  final String? action;

  final Set<UserAccessProfile> profiles;

  bool isVisibleTo(UserAccessProfile? profile) =>
      profiles.isEmpty || profiles.contains(profile);
}

class ModuleMenuItem {
  const ModuleMenuItem({
    required this.id,
    required this.label,
    required this.icon,
    required this.route,
    this.push = false,
    this.profiles = const {},
  });

  final String id;
  final String label;
  final AppIconData icon;

  /// Rota absoluta do destino.
  final String route;

  /// `true` quando o destino é uma tela funda com "Voltar" próprio (ex.: uma
  /// funcionalidade aberta direto do menu): o shell empilha a rota em vez de
  /// substituí-la, para o "Voltar" ter para onde retornar.
  final bool push;

  final Set<UserAccessProfile> profiles;

  bool isVisibleTo(UserAccessProfile? profile) =>
      profiles.isEmpty || profiles.contains(profile);
}

class ModuleMenuSection {
  const ModuleMenuSection({required this.title, required this.items});

  final String title;
  final List<ModuleMenuItem> items;
}

class ModuleDef {
  const ModuleDef({
    required this.id,
    required this.label,
    required this.icon,
    required this.homeRoute,
    required this.bottomTabs,
    this.menuSections,
    this.menuSectionsBuilder,
  });

  final String id;
  final String label;
  final AppIconData icon;

  /// Rota inicial absoluta do módulo.
  final String homeRoute;
  final List<BottomTab> bottomTabs;

  /// Funcionalidades exibidas no RevealMenu (aba Mais) — contextuais ao módulo.
  /// Quando ausente, o menu deriva uma seção única das bottomTabs navegáveis.
  final List<ModuleMenuSection>? menuSections;

  /// Alternativa a [menuSections] quando as seções derivam de dados em vez de
  /// uma lista fixa (ex.: grupos do catálogo funcional). Tem precedência.
  final List<ModuleMenuSection> Function()? menuSectionsBuilder;
}

/// Navegação primária da entrada operacional: duas abas de cada lado e, no
/// centro, o "+" de adição rápida ([quickAddAction]), que abre
/// [operationalQuickAdds] em vez de navegar.
///
/// Confinamento saiu da barra para dar lugar ao "+": o grupo continua no
/// menu lateral ([operationalMenuSections]) e as duas rotinas diárias dele
/// (trato e leitura de cocho) estão na adição rápida.
const List<BottomTab> operationalBottomTabs = [
  BottomTab(
    id: 'inicio',
    label: 'Início',
    icon: AppIcons.home,
    path: 'operacional',
  ),
  BottomTab(
    id: 'pecuaria',
    label: 'Pecuária',
    icon: AppIcons.pecuaria,
    path: 'operacional/grupo/pecuaria',
  ),
  BottomTab(
    id: 'adicionar',
    label: 'Adicionar',
    icon: AppIcons.plus,
    path: '',
    action: quickAddAction,
  ),
  BottomTab(
    id: 'agricultura',
    label: 'Agricultura',
    icon: AppIcons.agricultura,
    path: 'operacional/grupo/agricultura',
  ),
  BottomTab(
    id: 'menu',
    label: 'Menu',
    icon: AppIcons.menu,
    path: '',
    action: 'menu',
  ),
];

/// Ação do "+" central da navbar.
const String quickAddAction = 'quick-add';

/// Adição rápida (o "+" da navbar): as rotinas de lançamento mais usadas no
/// dia a dia de campo, no máximo cinco. O destino vem do catálogo funcional
/// pelo id — a mesma rota do menu e dos atalhos —, só o rótulo é mais curto.
List<ModuleMenuItem> operationalQuickAdds() {
  const atalhos = [
    ('apontamento', 'Apontamento', AppIcons.tractor),
    ('pesagem', 'Pesagem', AppIcons.scale),
    ('trato-diario', 'Trato diário', AppIcons.wheat),
    ('leitura-cocho-confinamento', 'Leitura de cocho', AppIcons.clipboardCheck),
    ('sanitario', 'Manejo sanitário', AppIcons.heartPulse),
  ];
  return [
    for (final (id, label, icon) in atalhos)
      ModuleMenuItem(
        id: id,
        label: label,
        icon: icon,
        route: operationalFeatureRoute(
          operationalFeatures.firstWhere((f) => f.id == id),
        ),
        push: true,
      ),
  ];
}

/// Itens do menu lateral do Operacional — e, sem o "Início", os atalhos da
/// tela inicial (`OperacionalHomeScreen`): um item por grupo do catálogo, na
/// ordem de produto. O menu é o índice completo das rotinas e repete de
/// propósito o que a navbar já mostra (Início, Pecuária, Agricultura).
///
/// "Ordens de serviço" vem logo depois do Início e abre a lista completa com
/// filtro (`/fazendas/campo/minhas-os`, empilhada), a mesma do "Ver todas".
List<ModuleMenuSection> operationalMenuSections() {
  const osGroup = 'Ordem de serviço';
  final entries = operationalGroupEntries();
  return [
    ModuleMenuSection(
      title: 'Menu',
      items: [
        const ModuleMenuItem(
          id: 'inicio-operacional',
          label: 'Início',
          icon: AppIcons.home,
          route: operationalHomeRoute,
        ),
        for (final entry in [
          ...entries.where((e) => e.group == osGroup),
          ...entries.where((e) => e.group != osGroup),
        ])
          ModuleMenuItem(
            id: groupToSlug(entry.group),
            label: entry.group == osGroup ? 'Ordens de serviço' : entry.label,
            icon: groupIcon(entry.group),
            route: entry.route,
            push: entry.isFeature,
          ),
      ],
    ),
  ];
}

/// Tela inicial do Operacional (aba "Início" da navbar).
const String operationalHomeRoute = '/fazendas/operacional';

/// Fallback do RevealMenu: seção única derivada das abas navegáveis do módulo.
List<ModuleMenuSection> getMenuSections(
  ModuleDef module, {
  UserAccessProfile? profile,
}) {
  final declared = module.menuSectionsBuilder?.call() ?? module.menuSections;
  if (declared != null) {
    return declared
        .map(
          (section) => ModuleMenuSection(
            title: section.title,
            items: section.items
                .where((item) => item.isVisibleTo(profile))
                .toList(),
          ),
        )
        .where((section) => section.items.isNotEmpty)
        .toList();
  }
  return [
    ModuleMenuSection(
      title: 'Funcionalidades',
      items: visibleBottomTabs(module, profile)
          .where((tab) => tab.path.isNotEmpty && tab.action == null)
          .map(
            (tab) => ModuleMenuItem(
              id: tab.id,
              label: tab.label,
              icon: tab.icon,
              route: '/${module.id}/${tab.path}',
            ),
          )
          .toList(),
    ),
  ];
}

List<BottomTab> visibleBottomTabs(
  ModuleDef module,
  UserAccessProfile? profile,
) => module.bottomTabs.where((tab) => tab.isVisibleTo(profile)).toList();

const List<ModuleDef> modules = [
  ModuleDef(
    id: 'fazendas',
    label: 'Fazendas',
    icon: AppIcons.sprout,
    homeRoute: '/fazendas',
    bottomTabs: [
      BottomTab(
        id: 'rotinas',
        label: 'Rotinas',
        icon: AppIcons.clipboardList,
        path: 'operacional',
      ),
      BottomTab(
        id: 'mais',
        label: 'Mais',
        icon: AppIcons.moreHorizontal,
        path: 'mais',
        action: 'menu',
      ),
    ],
    // Os grupos do catálogo operacional (antes a grade da tela inicial)
    // entram no grupo único "Menu" — ver [operationalMenuSections].
    menuSectionsBuilder: operationalMenuSections,
  ),
];

final Map<String, ModuleDef> moduleMap = {for (final m in modules) m.id: m};

ModuleDef? getModule(String? id) => id == null ? null : moduleMap[id];
